### PR TITLE
Migrate to React 16

### PR DESCRIPTION
**Pull request by mthuret**
Just to test with CI 

Remaining known issues: 

- [x] recipes test passes but there seems to have a "timeout" issue that make the script failed. For instance, when removing the content of the autocomplete tests, `npm run test:recipes` pass. 

- [x] storybook is failing like crazy. See: https://github.com/storybooks/storybook/issues/1965 (workaround found by deactivating animations on the sidebar).

- [ ] warning in test about Error Boundaries. See: https://github.com/facebook/react/issues/11098 (not mandatory)

### CODE REVIEW

File: packages/react-instantsearch/examples/autocomplete/package.json [hunk 1]
```diff
@@ -2,18 +2,19 @@
   "private": true,
   "devDependencies": {
     "identity-obj-proxy": "3.0.0",
+    "jest-environment-jsdom-11.0.0": "^20.0.9",
```
Comment (vvo) [inline packages/react-instantsearch/examples/autocomplete/package.json]: this is now an external package is that it?

File: packages/react-instantsearch/examples/autocomplete/package.json [hunk 2]
```diff
@@ -23,7 +24,9 @@
     "notify": false,
     "moduleNameMapper": {
       "\\.(css)$": "identity-obj-proxy"
-    }
+    },
+    "setupFiles": ["../../../../shim.js"],
```
Comment (vvo) [inline packages/react-instantsearch/examples/autocomplete/package.json]: Maybe you can use <ROOT_DIR> smg here?

File: shim.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// see: https://github.com/facebookincubator/create-react-app/issues/3199
```
Comment (vvo) [inline shim.js]: 👍  Love when we do that :)

**Comment by vvo** [comment]
That's some serious chore/build stuff, much needed to ensure we are always up with the latest versions! GG

File: packages/react-instantsearch/examples/autocomplete/package.json (hunk 1 above)
Comment (mthuret) [reply packages/react-instantsearch/examples/autocomplete/package.json]: http://facebook.github.io/jest/docs/en/configuration.html#testenvironment-string

File: packages/react-instantsearch/examples/autocomplete/package.json (hunk 2 above)
Comment (mthuret) [reply packages/react-instantsearch/examples/autocomplete/package.json]: Could be better indeed :)

Comment (mthuret) [reply packages/react-instantsearch/examples/autocomplete/package.json]: @vvo is <rootDir> here referring to the directory where the **example** package.json is, and therefore, can't work if shim.js is only at the very root of the project?

Comment (vvo) [reply packages/react-instantsearch/examples/autocomplete/package.json]: Indeed root_dir would be the example/package.json.

Also this seems temporary given comments in https://github.com/facebookincubator/create-react-app/issues/3199, just watch the issue and remove it afterwards

**Comment by vvo** [comment]
LGTM

File: shim.js (hunk 3 above)
Comment (Haroenv) [reply shim.js]: It's already fixed in Jest, just not released yet; see https://github.com/facebook/jest/issues/4545#issuecomment-332762365

**Pull request merged**